System/OsString/Internal.hs

```diff
@@ -172,3 +181,540 @@ toChar (OsChar (WindowsChar w)) = chr $ fromIntegral w
+--
+-- @since 1.4.200.0
+cons :: OsChar -> OsString -> OsString
+cons = coerce PF.cons
+
+-- | /O(1)/ Extract the last element of a OsString, which must be finite and non-empty.
+-- An exception will be thrown in the case of an empty OsString.
+--
+-- This is a partial function, consider using 'unsnoc' instead.
+--
+-- @since 1.4.200.0
+last :: HasCallStack => OsString -> OsChar
+last = coerce PF.last
+
+-- | /O(n)/ Extract the elements after the head of a OsString, which must be non-empty.
+-- An exception will be thrown in the case of an empty OsString.
+--
+-- This is a partial function, consider using 'uncons' instead.
+--
+-- @since 1.4.200.0
+tail :: HasCallStack => OsString -> OsString
+tail = coerce PF.tail
+
+-- | /O(n)/ Extract the 'head' and 'tail' of a OsString, returning 'Nothing'
+-- if it is empty.
+--
+-- @since 1.4.200.0
+uncons :: OsString -> Maybe (OsChar, OsString)
+uncons = coerce PF.uncons
+
+-- | /O(1)/ Extract the first element of a OsString, which must be non-empty.
+-- An exception will be thrown in the case of an empty OsString.
+--
+-- This is a partial function, consider using 'uncons' instead.
+--
+-- @since 1.4.200.0
+head :: HasCallStack => OsString -> OsChar
+head = coerce PF.head
+
+-- | /O(n)/ Return all the elements of a 'OsString' except the last one.
+-- An exception will be thrown in the case of an empty OsString.
+--
+-- This is a partial function, consider using 'unsnoc' instead.
+--
+-- @since 1.4.200.0
+init :: HasCallStack => OsString -> OsString
+init = coerce PF.init
+
+-- | /O(n)/ Extract the 'init' and 'last' of a OsString, returning 'Nothing'
+-- if it is empty.
+--
+-- @since 1.4.200.0
+unsnoc :: OsString -> Maybe (OsString, OsChar)
+unsnoc = coerce PF.unsnoc
+
+-- | /O(1)/ Test whether a 'OsString' is empty.
+--
+-- @since 1.4.200.0
+null :: OsString -> Bool
+null = coerce PF.null
+
+-- | /O(1)/ The length of a 'OsString'.
+--
+-- @since 1.4.200.0
+length :: OsString -> Int
+length = coerce PF.length
+
+-- | /O(n)/ 'map' @f xs@ is the OsString obtained by applying @f@ to each
+-- element of @xs@.
+--
+-- @since 1.4.200.0
+map :: (OsChar -> OsChar) -> OsString -> OsString
+map = coerce PF.map
+
+-- | /O(n)/ 'reverse' @xs@ efficiently returns the elements of @xs@ in reverse order.
+--
+-- @since 1.4.200.0
+reverse :: OsString -> OsString
+reverse = coerce PF.reverse
+
+-- | /O(n)/ The 'intercalate' function takes a 'OsString' and a list of
+-- 'OsString's and concatenates the list after interspersing the first
+-- argument between each element of the list.
+--
+-- @since 1.4.200.0
+intercalate :: OsString -> [OsString] -> OsString
+intercalate = coerce PF.intercalate
+
+-- | 'foldl', applied to a binary operator, a starting value (typically
+-- the left-identity of the operator), and a OsString, reduces the
+-- OsString using the binary operator, from left to right.
+--
+-- @since 1.4.200.0
+foldl :: (a -> OsChar -> a) -> a -> OsString -> a
+foldl f a (OsString s) = PF.foldl (coerce f) a s
```

Comment:
Can you apply the same `ScopedTypeVariables` `coerce`-disambiguation trick on the new polymorphic functions in this module?